gpii/node_modules/journal/src/Journal.js

```diff
@@ -0,0 +1,278 @@
+ *
+ * Copyright 2016 Raising the Floor - International
+ *
+ * Licensed under the New BSD license. You may not use this file except in
+ * compliance with this License.
+ *
+ */
+
+"use strict";
+
+var fluid = require("infusion"),
+    kettle = fluid.registerNamespace("kettle"),
+    gpii = fluid.registerNamespace("gpii"),
+    fs = require("fs"),
+    writeFileAtomic = require("write-file-atomic"),
+    glob = require("glob");
+
+/** Manages reading and writing of journal files held within the "GPII settings directory",
+ * managed in a suitable platform-specific area of the user's settings files.
+ */
+
+fluid.defaults("gpii.journal", {
+    gradeNames: "fluid.component",
+    maxOldJournals: 20,
+    components: {
+        settingsDir: {
+            type: "gpii.settingsDir"
+        }
+    },
+    events: {
+        onUpdateSnapshot: null
+    },
+    listeners: {
+        "onCreate.readJournals": "gpii.journal.readJournals({that})",
+        "onCreate.trimJournals": {
+            funcName: "gpii.journal.trimJournals",
+            args: "{that}",
+            priority: "after:readJournals"
+        },
+        "onCreate.findCrashed": {
+            funcName: "gpii.journal.findCrashed",
+            args: "{that}",
+            priority: "after:trimJournals"
+        }
+    },
+    invokers: {
+        // args: {that}, {session}, {originalSettings}, closed
+        writeJournal: "gpii.journal.writeJournal"
+    }
+});
+
+/** A mixin grade supplied to gpii.journal accounting for its interaction with a lifecycleManager.
+ * In the main architecture this is supplied with the journal's definition in gpii.flowManager.local,
+ * which also incidentally assures that this component will definitely be constructed after the
+ * lifecycleManager.
+ */
+
+fluid.defaults("gpii.journalLifecycleManager", {
+    gradeNames: "fluid.component",
+    listeners: {
+        "{lifecycleManager}.events.onSessionSnapshotUpdate": {
+            namespace: "writeJournal",
+            func: "{that}.writeJournal",
+            args: ["{that}", "{arguments}.1", "{arguments}.1.model.originalSettings", false]
+        },
+        "{lifecycleManager}.events.onSessionStop": {
+            namespace: "writeJournal",
+            func: "{that}.writeJournal",
+            args: ["{that}", "{arguments}.1", "{arguments}.1.model.originalSettings", true]
+        }
+    },
+    invokers: {
+        restoreJournal: {
+            funcName: "gpii.journal.restoreJournal",
+            args: ["{that}", "{lifecycleManager}", "{arguments}.0"] // journalId
+        }
+    }
+});
+
+
+gpii.journal.restoreJournal = function (journal, lifecycleManager, journalId) {
+    var parsed = gpii.journal.parseId(journalId);
+    if (parsed.isError) {
+        fluid.fail(parsed.message);
+    } else {
+        fluid.log("Restoring journal with id " + journalId);
+        var journal = gpii.journal.fetchJournal(journal.journalFiles, parsed);
```

Comment:
Looks like this is covered by the following rule: http://eslint.org/docs/rules/no-redeclare which we should similarly add